### PR TITLE
DEV-1074: REST API for JSON DIDDocs

### PR DESCRIPTION
**Pull request by Toktar**

### CODE REVIEW

**Comment by ankurdotb** [comment]
Task linked: [DEV-957 Implement API endpoint to query DIDs and fetch properly formatted full DIDDocs](https://app.clickup.com/t/6600954/DEV-957)

File: proto/cheqd/v1/query.proto [hunk 1]
```diff
@@ -13,6 +13,9 @@ service Query {
 	rpc Did(QueryGetDidRequest) returns (QueryGetDidResponse) {
 		option (google.api.http).get = "/cheqd/cheqdnode/cheqd/did/{id}";
 	}
+	rpc DidJson(QueryGetDidJsonRequest) returns (QueryGetDidJsonResponse) {
+		option (google.api.http).get = "/cheqd/cheqdnode/cheqd/did+json/{id}";
```
Comment (ankurdotb) [inline proto/cheqd/v1/query.proto:17]: Why is the URL path to fetch this so weird? Added a comment over on ClickUp for extended discussion: https://app.clickup.com/t/6600954/DEV-957?comment=560490023

File: proto/cheqd/v1/did.proto [hunk 2]
```diff
@@ -7,7 +7,7 @@ import "google/protobuf/any.proto";
 import "cheqd/v1/common.proto";
 
 message Did {
-  repeated string context = 1; // optional
+  repeated string context = 1 [json_name="@context"]; // optional
   string id = 2;
   repeated string controller = 3; //optional
   repeated VerificationMethod verification_method = 4; // optional
```
Comment (ankurdotb) [inline proto/cheqd/v1/did.proto:13]: Just like the verification method, does it not make sense to have the following ones (L14-18) also as an object? These are all a set of Verification Method maps: https://w3c.github.io/did-core/#did-document-properties

**Review by ankurdotb — CHANGES_REQUESTED** [review]
At the very least, this requires a discussion on URI paths for the API (and what the responses are)

File: proto/cheqd/v1/did.proto (hunk 2 above)
Comment (ankurdotb) [reply proto/cheqd/v1/did.proto:13]: Also, this is likely going to tougher than what _this_ PR can achieve, so one to consider for the next sprint.

Comment (Toktar) [reply proto/cheqd/v1/did.proto:13]: I'm not sure that I correctly understood this idea. Let's discuss it on a call

**Review by askolesov — APPROVED** [review]
Ok, this implementation is good as a temporary solution. The only issue that I see is: `ver_method -> public_key_jwk` is a map in the did core spec but in our proto definitions it's an array and there is no way to make it map. So most probably we will have to introduce the second layer of models to convert our representation to did core spec-compliant one.

File: x/cheqd/keeper/grpc_query_did.go [hunk 3]
```diff
@@ -27,3 +33,29 @@ func (k Keeper) Did(c context.Context, req *types.QueryGetDidRequest) (*types.Qu
...
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	stateValue, err := k.GetDid(&ctx, req.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	did, err := stateValue.UnpackDataAsDid()
```
[the remark this answers is not in the document]
Comment (askolesov) [reply x/cheqd/keeper/grpc_query_did.go:49]: As in other endpoints

[the remark this answers is not in the document]
Comment (askolesov) [reply x/cheqd/keeper/grpc_query_did.go:49]: Oh, I foresee big challenges here...

**Review by askolesov — CHANGES_REQUESTED** [review]
Sorry, I noticed that metadata is missing...

File: x/cheqd/keeper/grpc_query_did.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ankurdotb) [reply x/cheqd/keeper/grpc_query_did.go:49]: Yeah had a long chat with Renata, I think we can/should split cheqd functionality into modules under cheqd (a bit like Cosmos SDK does), which will allow us to bump/fix them are different rates, not changing the consensus version of one big cheqd module blob. (Worth discussing how feasible this idea is.)

[the remark this answers is not in the document]
Comment (askolesov) [reply x/cheqd/keeper/grpc_query_did.go:49]: Queries don't bump consensus version, only changes in transactions. I would also like to take a look at the proposed design.

[the remark this answers is not in the document]
Comment (askolesov) [reply x/cheqd/keeper/grpc_query_did.go:49]: > BTW when the DID Core spec says it's a "map", they don't necessarily mean a "map" as in a native map object/format. https://infra.spec.whatwg.org/#ordered-map

I think they speak in Js terms where maps and objects are the same.

As for our implementation, usage of maps vs objects depends on whether we want to allow extension of verification methods or not.

**Pull request closed without merging**